Velocity-CTD Rewrite

R00tB33rMan opened the pull request:
The long-awaited rewrite of CTD that stabilizes and improves the entirety of CTD.

xCodiq committed:
Progress push

xCodiq committed:
Merge remote-tracking branch 'origin/xcd' into xcd

# Conflicts:
#	proxy/src/main/java/com/velocitypowered/proxy/redis/RedisManagerImpl.java
#	proxy/src/main/java/com/velocitypowered/proxy/redis/multiproxy/MultiProxyHandler.java

xCodiq committed:
Merge remote-tracking branch 'origin/libdeflate' into xcd

# Conflicts:
#	proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java

R00tB33rMan committed:
Upstream

xCodiq committed:
feat(xcd): more work done towards a final product.

Commit didn't pull upstream yet.

xCodiq committed:
Merge remote-tracking branch 'origin/xcd' into xcd

xCodiq committed:
feat(xcd): even more work!

xCodiq committed:
feat(xcd): more work; without doing a checkstyle session

xCodiq committed:
chore(xcd): passed spotlessJavaCheck

xCodiq committed:
chore(xcd): started working on checkstyle and javadocs

xCodiq committed:
chore(xcd): done checkstyle, javadocs and spotless java modifications

xCodiq committed:
fix(xcd): invert player online check in QueueAdminCommand

xCodiq committed:
fix(xcd): streamline player transfer and cleanup in VelocityServer

xCodiq committed:
feat(xcd): implement teardown method for resource cleanup in queue management & use unique id's instead of Player object in player related queue-operations

xCodiq committed:
fix(xcd): change log level from info to debug for packet publication success

xCodiq committed:
fix(xcd): continued work on xcd

Commit:
chore(xcd): changed key from String/username to UUID/holderUuid

Commit:
chore(xcd): cleanup todos

xCodiq committed:
chore(xcd): cleanup and optimize imports

xCodiq committed:
chore(xcd): added missing copyright

xCodiq committed:
chore(xcd): changed final adventure arguments approach for redis related components

xCodiq committed:
chore(xcd): minor cleanups

xCodiq committed:
Merge remote-tracking branch 'origin/libdeflate' into xcd-merge

# Conflicts:
#	proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java
#	proxy/src/main/java/com/velocitypowered/proxy/redis/multiproxy/MultiProxyHandler.java

xCodiq committed:
chore(xcd): fix compile and checkstyle errors caused by the merge

xCodiq committed:
feat(xcd): implemented an option to configure auto-queue servers

xCodiq committed:
fix(xcd): reworked the auto queue feature

R00tB33rMan committed:
Cleanup rewrite to fully match codestyle

R00tB33rMan committed:
Use existing element for auto queue tray

R00tB33rMan committed:
Base Branch Modifications

R00tB33rMan committed:
Cleanup logger references and ensure static fields utilize correct parameterization

R00tB33rMan committed:
Oopsie

R00tB33rMan committed:
Upstream

R00tB33rMan committed:
Upstream

R00tB33rMan committed:
Merge branch 'libdeflate' into xcd-merge

xCodiq committed:
fix(xcd): ensure player removal from queue only if queue is enabled

R00tB33rMan committed:
Upstream

R00tB33rMan committed:
Merge

R00tB33rMan committed:
Upstream

R00tB33rMan committed:
Upstream

R00tB33rMan committed:
Upstream

R00tB33rMan committed:
Upstream

R00tB33rMan committed:
Upstream

WouterGritter commented on proxy/src/main/java/com/velocitypowered/proxy/VelocityServer.java:
```diff
@@ -327,19 +326,15 @@ public class VelocityServer implements ProxyServer, ForwardingAudience {
+   * Coordinates server queues and handles queue assignment logic.
    */
-  private RedisManagerImpl redisManager;
+  private com.velocitypowered.proxy.queue.manager.QueueManager<?> queueManager;
```
weird class reference doesn't have to be FQCN

WouterGritter commented:
Numerous implementations are written twice, once where a redis server is used, and another time for a "local" implementation.
`VelocityAlert` for example:
In `AlertCommand`:
```java
if (server.isRedisEnabled()) {
  new VelocityAlert(alertComponent).publish();
} else {
  server.sendMessage(alertComponent); // implementation
}
```

In `RouteRegistry`:
```java
VELOCITY_ALERT(VelocityAlert.class, PacketBehaviour.SEND_COMPONENT::behave),
```

Links to `PacketBehaviour`:
```java
PacketBehaviour<Audience, ComponentPacket> SEND_COMPONENT = (carrier, packet) -> {
  final Component component = packet.deserialize();
  if (component != null) {
    carrier.sendMessage(component); // implementation
  }
};
```

Ideally we generate the `VelocityAlert` packet once, and have something upstream decide if we need to forward this to a redis server, or if we can execute it locally.
`VelocityServer.isRedisEnabled()` is called in 46 places, so it might be worth looking into similar patterns of duplicate code.

Other packets like `VelocityActionBar` seem to be only used internally when redis is enabled. These things should also be handled by something upstream, such that we can send packets like VelocityActionBar without having to know if we're connected with a redis server or not.

Note: if redis is enabled, is it still possible to run in a single-proxy environment? If this is the case, the implementation might look very simple, just always generate a packet and push it to the "redis" server, which could just be a dummy implementation that bridges packets directly onto any listeners. Wouldn't require much code, something to look into.

I see the same thing happening a lot. `FindCommand` also has two implementations for finding players, one with redis enabled and one without. We should really have all of these be the same flow:

```java
interface PlayerDepotService {
  PlayerEntry getPlayerEntry(UUID uuid);
}

class LocalPlayerDepotService {
	// no redis connection, local state only
}

class RedisPlayerDepotService {
	// connect to redis, use that for the state
}
```

This way consumers of this data, e.g. the commands, really don't (and shouldn't) care if we use no DB, redis, or anything else in the future. They will only interface with `PlayerDepotService`, which would either be a local or a redis implementation.

WouterGritter commented:
Large amount of comments
I like to write code with the least amount of comments possible, code should be self-explainatory when reading it. Comments will get stale/out of date when code moves/changes and (imo) adds a lot of clutter when I just want to read the code.

Take the register() method block comment for each command:
```java
/**
 * Returns the command instance if enabled, or {@code null} if disabled via configuration.
 *
 * @return the command instance or {@code null} if disabled
 */
```

This is repeated in each command. If something changes, you'd have to modify each comment, which we're definitely going to forget. I agree that such a comment can be helpful, but it would be better to make an interface `RegistrableCommand` with a `register()` method with this comment, and just implement it without a comment in each class.

Same thing for a comment on a field in `AlertCommand`:
```java
/**
 * The server instance used to perform broadcasts and access multi-proxy configuration.
 */
private final VelocityServer server;
```
Sure, it gets used for broadcasts and accecing configuration. But if we use the `server` field for anything else in this class the comment is wrong and outdated. And it doesn't really matter what we use it for. Such a field really doesn't need a comment, and just makes it so I only see a handful of actual lines of code at the top of this class.

There are many more comments like this. I think it could be really helpful if we remove all of the comments that repeat exactly what a block of code does, comments that don't add any useful information and comments that would get out of date easily.

WouterGritter committed:
Upstream

WouterGritter committed:
Suppress warning for implementing a deprecated method

WouterGritter committed:
Clean up field definition

WouterGritter committed:
Build (but not release) xcd-merge on every push to github

WouterGritter commented:
b4e2f10d (upstream merge on 26 jan) broke this branch, as there were no changes in between other than upstream merges, I have forced this branch to the upstream commit before the broken one and merged upstream correctly again.

WouterGritter committed:
Merge branch 'libdeflate' into xcd-merge

WouterGritter committed:
Lenient comment checkstyle for velocity-proxy (implementation) module.

WouterGritter committed:
Remove comment clutter for commands

WouterGritter committed:
Don't use records for commands

WouterGritter committed:
Fix checkstyle

WouterGritter committed:
Command building/registration refactor

WouterGritter committed:
Merge branch 'libdeflate' into xcd-merge

WouterGritter committed:
Fix /<server_name> by using /server <server_name>'s implementation

WouterGritter committed:
Missing @Override for build() in commands

WouterGritter committed:
Fix /hub by using same implementation as /server <server_name> and /<server_name> by moving player send/queue logic to VelocityCommands

WouterGritter committed:
Fix /send command problems by completely reimplementing the command

WouterGritter committed:
Make dynamic-proxy-filter an enum

WouterGritter committed:
Make dynamic-fallbacks-filter an enum

WouterGritter committed:
Align config defaults for new configurable enums to existing style (lower case)

WouterGritter committed:
Merge branch 'libdeflate' into xcd-merge

WouterGritter committed:
Merge branch 'libdeflate' into xcd-merge

WouterGritter committed:
Fix toLowerCase locale issue

WouterGritter committed:
Uppercase LOGGER fields

WouterGritter committed:
Move calculating `serversToTry` to utility class to avoid duplicate code

WouterGritter committed:
Implement ServerRetrySession to fix Velocity-CTD implementation of getNextServerToTry() (broke due to dynamic ordering of fallback servers)

WouterGritter committed:
Refactor ForcedHostResolver#resolveServersToTry to make the logic more clear

WouterGritter committed:
Helper class naming

WouterGritter committed:
Remove duplicate string declarations for "velocity:callback"

WouterGritter committed:
Don't call setServerLinks twice on connect (redundant). Fixes calling getNextServerToTry twice on login which messes up the retry session state

WouterGritter committed:
Introduce configurable `forced-host-as-fallback`

WouterGritter committed:
Use FallbackServerResolver#resolveServersToTry in /hub to stay consistent with ServerRetrySession#getNextServerToTry

WouterGritter committed:
Fix spotless violations

WouterGritter committed:
Fix auto queueing

WouterGritter committed:
Fix checkstyle

WouterGritter committed:
Allow queueing to multiple servers with `auto-queue-servers` if multi queueing is enabled

WouterGritter committed:
More complex permission check for ProxyAliasCommand to avoid tab suggestions when the player doesn't pass the requirement checks of the underlying alias command(s)

WouterGritter committed:
Only return queue command aliases for /server when queue is enabled

R00tB33rMan committed:
Upstream

WouterGritter committed:
Small refactor of the Queue interface

WouterGritter committed:
Merge remote-tracking branch 'origin/xcd-merge' into xcd-merge

WouterGritter committed:
Fix checkstyle

WouterGritter committed:
- fix full server error message
- fix full server join spam
- fix velocity.queue.full.bypass permission (allow joining full servers)
- remove unreachable server ping logic

WouterGritter committed:
Upstream

WouterGritter committed:
Don't debug log on redis packet publish success

WouterGritter committed:
Transfer player at most once per transfer tick

WouterGritter committed:
Early exit for AbstractQueueManager#onPlayerDisconnect to avoid calling getTimeoutInSeconds when proxy is shutting down

WouterGritter committed:
Fix `/glist [all]` when proxy should be proxies (plural)

WouterGritter committed:
Refactor duplicate code by adding "magic server/proxy" argument (e.g. "all") to VelocityCommands#suggestServer and VelocityCommands#suggestProxy

WouterGritter committed:
Align /plist behavior to /glist, including allowing "all" for proxy and server (e.g. /plist all all)

WouterGritter committed:
Merge branch 'libdeflate' into xcd-merge

WouterGritter committed:
Resolve merge conflicts from previous merge

WouterGritter committed:
Fix actionbar message for multi queue

WouterGritter committed:
Remove arbitrary queue limits

WouterGritter committed:
Implement advanced permission framework

WouterGritter committed:
Implement luckperms permission integration (through the new AdvancedPermissionResolver api)

WouterGritter committed:
Make `AdvancedPermissionResolver#getPermissionMap` `@Nullable`

WouterGritter committed:
Improve performance for timeout and priority permission checks by making use of new permission api

WouterGritter committed:
Fix queue ETA message

WouterGritter committed:
Use RFC 5737 TEST-NET addresses for default public IPs of the proxies

WouterGritter committed:
Merge branch 'libdeflate' into xcd-merge

WouterGritter committed:
Log warning message when slow permission check is used

WouterGritter committed:
Add unused ConnectedPlayer#setPermissionFunction back in some external plugins rely on

WouterGritter committed:
Merge branch 'libdeflate' into xcd-merge

WouterGritter committed:
Resolve conflict after merge

WouterGritter committed:
Fix logic bugs where:
- a transfer tick wouldn't transfer anyone if the player at position 1 had already transferred in the same tick
- a queue would never transfer a player with isFullBypass() if the server was full, and they weren't number 1

WouterGritter committed:
Rename methods to make intent clear and fix bug where the first player gets removed instead of an intended player (possibly not the first one)

WouterGritter committed:
Abort transfer when needed to avoid stuck `waitingForConnection = true` on a race condition (causing the queue to be stuck)

WouterGritter committed:
Implement interface layer for new queue system

WouterGritter committed:
Delete old queue classes (doesn't compile)

WouterGritter committed:
Implement first iteration of new queue system

WouterGritter committed:
Update callers/users of new queue system

WouterGritter committed:
Split redis logic into RedisVelocityQueue[Entry|Manager] and slight queue system refactor

WouterGritter committed:
Only call `PlayerDepotService#onPlayerDisconnect` and `VelocityQueueManager#onPlayerDisconnect` if the player has been fully initialized

WouterGritter committed:
Implement QueuePlayerList that gives better time-complexity for reads and ensures thread safety

WouterGritter committed:
Merge branch 'libdeflate' into xcd-merge

WouterGritter committed:
Merge branch 'xcd-merge' into xcd-merge-queue-reimplementation

WouterGritter committed:
Use Component#translatable with translation key

WouterGritter commented on proxy/src/main/java/com/velocitypowered/proxy/command/builtin/HubCommand.java:
```diff
@@ -24,110 +24,82 @@
-      return 0;
-    }
+    ServerConnection con = player.getCurrentServer().orElse(null);
+    Objects.requireNonNull(con);
```
not fixing, an error state may throw an exception

WouterGritter commented on proxy/src/main/java/com/velocityctd/proxy/commands/builtin/SlashServerCommand.java:
```diff
@@ -26,77 +26,55 @@
-    if (conn != null && conn.getServerInfo().getName().equalsIgnoreCase(server.getServerInfo().getName())) {
+    ServerConnection connection = player.getCurrentServer().orElse(null);
+    if (connection != null && connection.getServer() == registeredServer) {
       player.sendMessage(Component.translatable("velocity.command.slashserver.already-connected"));
-      return Command.SINGLE_SUCCESS;
+      return -1;
     }
```
not fixing, works fine

WouterGritter commented on proxy/src/main/java/com/velocitypowered/proxy/redis/impl/transaction/VelocityTransaction.java:
```diff
@@ -0,0 +1,63 @@
+    super(sentPacket);
+
+    if (source != null && timeoutTranslationKey != null) {
+      this.onTimeout(t -> source.sendMessage(Component.text(timeoutTranslationKey, NamedTextColor.RED)));
```
fixed

WouterGritter commented on proxy/src/main/java/com/velocitypowered/proxy/protocol/netty/GameSpyQueryHandler.java:
```diff
@@ -132,19 +132,11 @@ public GameSpyQueryHandler(final VelocityServer server) {
         .gameVersion(ProtocolVersion.SUPPORTED_VERSION_STRING)
         .map(server.getConfiguration().getQueryMap())
-        .currentPlayers(online)
+        .currentPlayers(server.getPlayerCount())
```
works as intended, ai talking shit

WouterGritter commented on proxy/src/main/java/com/velocitypowered/proxy/connection/util/ServerListPingHandler.java:
```diff
@@ -113,20 +110,13 @@ private ServerPing constructLocalPing(ProtocolVersion version) {
     }
 
     return new ServerPing(
         new ServerPing.Version(version.getProtocol(), formatVersionString(serverPingVersion, version)),
-        new ServerPing.Players(online, configuration.getShowMaxPlayers(), samplePlayers),
+        new ServerPing.Players(server.getPlayerCount(), configuration.getShowMaxPlayers(), samplePlayers),
```
works as intended, ai talking shit

WouterGritter commented on proxy/src/main/java/com/velocitypowered/proxy/config/DynamicProxyFilterMode.java:
```diff
@@ -15,39 +15,36 @@
+   * proxy from the "master-proxy-ids" list.
    */
-  ServerQueueStatus get(String serverName);
+  FIRST_FOUND,
```
not fixing as this replaces strings with the exact same names in the `libdeflate` branch. as intended.

WouterGritter commented on proxy/src/main/java/com/velocitypowered/proxy/config/DynamicProxyFilterMode.java:
```diff
@@ -15,39 +15,36 @@
+   * Sends the player to the fallback proxy
+   * with the least number of players.
    */
+  MOST_EMPTY,
```
? same comment as above
not fixing as this replaces strings with the exact same names in the libdeflate branch. as intended.

WouterGritter commented on proxy/src/main/java/com/velocitypowered/proxy/connection/backend/BungeeCordMessageResponder.java:
```diff
@@ -173,26 +171,15 @@ private void processPlayerCount(final ByteBufDataInput in) {
-          amount = proxy.getPlayerCount();
-        }
-        out.writeInt(amount);
+        out.writeInt(proxy.getPlayerCount());
```
not true. third time stupid ai makes this mistake. `VelocityProxy#getPlayerCount` now gives the total player count when redis is enabled.

WouterGritter commented on proxy/src/main/java/com/velocityctd/proxy/redis/depot/Depot.java:
```diff
@@ -0,0 +1,90 @@
+ * @param <K> the key type used to index values in the depot
+ * @param <V> the value type stored in the depot, extending {@link DepotEntry}
+ */
+public interface Depot<K, V extends DepotEntry<K, V>> {
```
ai misunderstands implementation

WouterGritter commented on proxy/src/main/java/com/velocityctd/proxy/redis/depot/Depot.java:
```diff
@@ -0,0 +1,90 @@
+   *
+   * @return a collection of all keys in the depot
+   */
+  Collection<String> keys();
```
duplicate comment, microslop's ai is making a fool of itself

WouterGritter committed:
Annotate nullables

WouterGritter commented on proxy/src/main/java/com/velocitypowered/proxy/redis/transaction/TransactionHandler.java:
```diff
@@ -0,0 +1,91 @@
+   * @param packet the incoming packet to handle
+   * @return a newly created reply packet, or {@code null} if no reply is required
+   */
+  public abstract R handlePacket(T packet);
```
fixed

WouterGritter commented on proxy/src/main/java/com/velocityctd/proxy/redis/transaction/TransactionHandler.java:
```diff
@@ -0,0 +1,91 @@
+   * @param redisPacket the packet to create the reply packet from
+   * @return {@code null} if no reply packet is needed, otherwise the reply packet
+   */
+  public @Nullable R getReplyPacket(final @NotNull RedisPacket redisPacket) {
+    // noinspection unchecked
+    final R replyPacket = this.handlePacket((T) redisPacket);
+    if (replyPacket == null) {
+      return null;
+    }
```
already fixed, duplicate ai comment from microslop slopilot

WouterGritter committed:
Javadoc typo

WouterGritter commented on proxy/src/main/java/com/velocitypowered/proxy/redis/packet/RedisPacket.java:
```diff
@@ -0,0 +1,99 @@
+  void setReply(boolean reply);
+
+  /**
+   * Get the unique if of the {@link com.velocitypowered.proxy.redis.transaction.Transaction} this
```
fixed

WouterGritter commented on proxy/src/main/java/com/velocitypowered/proxy/config/DynamicProxyFilterMode.java:
```diff
@@ -15,39 +15,36 @@
+   * Sends the player to the fallback proxy
+   * with the most number of players.
+   */
+  LEAST_EMPTY,
```
duplicate comment from microslop ai

WouterGritter commented on proxy/src/main/java/com/velocitypowered/proxy/config/DynamicFallbackFilter.java:
```diff
@@ -0,0 +1,43 @@
+   * Sends the player to the fallback server
+   * with the most number of players.
+   */
+  MOST_POPULATED,
```
lol ai thinks this wont compile, check the pipeline. not fixing.

WouterGritter commented on proxy/src/main/java/com/velocityctd/proxy/commands/builtin/ProxyAliasCommand.java:
```diff
@@ -29,57 +32,89 @@
+  private Optional<Boolean> passesCommandRequirements(@NonNull CommandSource commandSource) {
+    VelocityCommandManager commandManager = (VelocityCommandManager) server.getCommandManager();
+    for (String command : commands) {
+      String commandRoot = command.split(" ", 2)[0];
+      CommandNode<CommandSource> commandNode = commandManager.getDispatcher().getRoot().getChild(commandRoot);
+      if (commandNode == null) {
+        return Optional.empty();
+      }
```
yes, nice scenario, but it wont ever be configured that way. all other command alias handling doesn't do this and the default config suggests you shouldn't add the leading `/`. not fixing.

WouterGritter commented on proxy/src/main/java/com/velocitypowered/proxy/config/DynamicProxyFilterMode.java:
```diff
@@ -15,39 +15,36 @@
+   * of the network and is not sent anywhere.
    */
-  List<ServerQueueStatus> getAll();
+  NONE,
```
🤡 microslop

WouterGritter commented:
The fixed Microslop copilot's suggestions have been implemented in https://github.com/GemstoneGG/Velocity-CTD/tree/xcd-merge-queue-reimplementation, this will get merged into this branch/PR soon (once the queue reimplementation is mature enough).

Microslop copilot suggested 1 actual good fix for an edge case which we would never hit, one real typo in a javadoc comment, and suggested to add `@Nullable` to an api which was correct.

WouterGritter committed:
Fix `/velocity sudo` losing command args when using redis

WouterGritter committed:
Fix requireNonNull checking wrong field

WouterGritter committed:
Fix NPE on reload when queue is disabled

WouterGritter committed:
Ensure a player is never auto-queued once they have left

WouterGritter commented:
> [xcd-merge-queue-reimplementation] will get merged into this branch/PR soon (once the queue reimplementation is mature enough).

Merged.

WouterGritter committed:
Update queue API to make ProxyServer#getQueueManager non-nullable

WouterGritter committed:
Fix checkstyle/spotless

WouterGritter committed:
Fix queue timeouts by not calling removePlayerEntirely() twice on server switch, and by not calling removePlayerEntirely() when connecting to the first server after joining the proxy

WouterGritter committed:
Implement migration to add default `[queue.auto-queue-servers]` config block

WouterGritter committed:
Add back in missing aliases for /shutdown command

WouterGritter committed:
Merge branch 'libdeflate' into xcd-merge

WouterGritter committed:
Implement `/gkick <player> [reason]` command

WouterGritter committed:
Implement `/ip <player>` command

WouterGritter committed:
Fix checkstyle issues

WouterGritter committed:
Optimize null checks

WouterGritter committed:
Rename `/ip` command to `/gip`

WouterGritter committed:
Fixes:
- Ghost players not tracked by proxy but still being connected when TransitionSessionHandler#disconnected got called during a transfer.
- Properly fix handling an error state during "Reconfiguring..." in ConfigSessionHandler#disconnected by completing the future NON-exceptionally with an error result.

Upstream's solution is (still) to throw an IOException here, causing in error spam and the wrong code being executed in handlers of the futures. Our solution used to be removing the player state (and hoping the connection would drop) but the connection could stay alive, introducing the ghost player. This new fix is to just handle it like an INTERNAL_SERVER_CONNECTION_ERROR, meaning the right choice will be made in the handler of the futures (either disconnect the player or abort the transfer). See https://github.com/PaperMC/Velocity/issues/1251

WouterGritter committed:
Use implementation classes instead of api interfaces in the implementation module (reduces unnecessary & excessive class casts)

WouterGritter committed:
Merge branch 'libdeflate' into xcd-merge

WouterGritter committed:
Remove scheduled queue removals for players with timeouts when they rejoin

WouterGritter committed:
Implement migrations to migrate upstream's config to a CTD config

WouterGritter committed:
Merge pull request #795 from GemstoneGG/ctd-config-migrations

Implement migrations to migrate upstream's config to a CTD config

WouterGritter committed:
Move all CTD-specific classes to com.velocityctd.[api/proxy].*

WouterGritter committed:
Split out (un)registerTranslations to utility class to prepare for messages.properties migrations

WouterGritter committed:
Implement migrating of files in `lang/` (e.g. `lang/messages.properties`)

WouterGritter committed:
Don't log package name for loggers in `com.velocityctd.*`

WouterGritter committed:
Implement proxy heartbeats on redis and reap servers when heartbeat expires

WouterGritter committed:
Merge branch 'libdeflate' into xcd-merge

R00tB33rMan commented:
We plan on merging this PR this upcoming weekend/week!

WouterGritter committed:
Don't throw when a chat packet comes in while the player isn't connected to a server

WouterGritter committed:
Shutdown redis & queue as late as possible to avoid exceptions on shutdown

WouterGritter committed:
Ping backend immediately after failed transfer attempt, to set the queue's server status to OFFLINE when appropriate

WouterGritter committed:
Fix inserting whitespaces before start of comment lines

WouterGritter committed:
Fix `kick-existing-players` and `kick-existing-players-check-ip`

WouterGritter committed:
Add missing migration for `kick-existing-players-check-ip`

WouterGritter committed:
Merge branch 'libdeflate' into xcd-merge

R00tB33rMan committed:
Adjust branch compilation iteration
